tests/state_test_util.go

```diff
@@ -198,63 +255,77 @@ func (t *StateTest) Run(subtest StateSubtest, vmconfig vm.Config, isTestExecutio
 	blockchain.InitDeriveSha(config)
 	block := t.genesis(config).ToBlock(common.Hash{}, nil)
 	memDBManager := database.NewMemoryDBManager()
-	statedb := MakePreState(memDBManager, t.json.Pre)
+	st = MakePreState(memDBManager, t.json.Pre)
 
 	post := t.json.Post[subtest.Fork][subtest.Index]
 	rules := config.Rules(block.Number())
 	msg, err := t.json.Tx.toMessage(post, rules, isTestExecutionSpecState)
 	if err != nil {
-		return nil, err
+		return st, common.Hash{}, err
 	}
+
+	// Try to recover tx with current signer
+	if len(post.TxBytes) != 0 {
+		var ttx types.Transaction
+		txBytes := post.TxBytes
+		if post.TxBytes[0] <= 4 {
+			txBytes = append([]byte{byte(types.EthereumTxTypeEnvelope)}, txBytes...)
+		}
+		err = ttx.UnmarshalBinary(txBytes)
+		if err != nil {
+			return st, common.Hash{}, err
+		}
+		if _, err := types.Sender(types.LatestSigner(config), &ttx); err != nil {
+			return st, common.Hash{}, err
+		}
+	}
+
 	txContext := blockchain.NewEVMTxContext(msg, block.Header(), config)
 	if isTestExecutionSpecState {
 		txContext.GasPrice, err = useEthGasPrice(rules, &t.json)
 		if err != nil {
-			return nil, err
+			return st, common.Hash{}, err
 		}
 	}
 	blockContext := blockchain.NewEVMBlockContext(block.Header(), nil, &t.json.Env.Coinbase)
 	blockContext.GetHash = vmTestBlockHash
-	evm := vm.NewEVM(blockContext, txContext, statedb, config, &vmconfig)
+	if isTestExecutionSpecState {
+		blockContext.GasLimit = t.json.Env.GasLimit
+	}
+	evm := vm.NewEVM(blockContext, txContext, st, config, &vmconfig)
 
 	if isTestExecutionSpecState {
 		useEthOpCodeGas(rules, evm)
 	}
 
-	snapshot := statedb.Snapshot()
+	snapshot := st.Snapshot()
 	result, err := blockchain.ApplyMessage(evm, msg)
 	if err != nil {
-		statedb.RevertToSnapshot(snapshot)
+		st.RevertToSnapshot(snapshot)
 	}
 
 	if isTestExecutionSpecState && err == nil {
-		useEthMiningReward(statedb, evm, &t.json.Tx, t.json.Env.BaseFee, result.UsedGas, txContext.GasPrice)
+		useEthMiningReward(st, evm, &t.json.Tx, t.json.Env.BaseFee, result.UsedGas, txContext.GasPrice)
 	}
 
-	if logs := rlpHash(statedb.Logs()); logs != common.Hash(post.Logs) {
-		return statedb, fmt.Errorf("post state logs hash mismatch: got %x, want %x", logs, post.Logs)
-	}
-
-	root, _ := statedb.Commit(true)
+	root, _ = st.Commit(true)
 	// Add 0-value mining reward. This only makes a difference in the cases
 	// where
 	// - the coinbase self-destructed, or
 	// - there are only 'bad' transactions, which aren't executed. In those cases,
 	//   the coinbase gets no txfee, so isn't created, and thus needs to be touched
-	statedb.AddBalance(block.Rewardbase(), new(big.Int))
+	st.AddBalance(block.Rewardbase(), new(big.Int))
 	// And _now_ get the state root
-	root = statedb.IntermediateRoot(true)
+	root = st.IntermediateRoot(true)
 
-	if isTestExecutionSpecState {
-		root, err = useEthStateRoot(statedb)
+	if err == nil && isTestExecutionSpecState {
```

Comment:
Is this err check needed?

Reply:
useEthStateRoot can overwrite non-nil error to nil.

Reply:
Isn't this error checking enough?
https://github.com/kaiachain/kaia/pull/149/files/870a8b4894fbaa626fb1c4b51862eeb38b8ddf73#diff-0f2a03d410f3b42fb5cc232db0be82b42a741b952aa3999887370ef91c1eca58R323-R325

```suggestion
	if isTestExecutionSpecState {
```

Reply:
Yes. If useEthStateRoot overwrite non-nil error to nil, It can't capture the error.

Reply:
Can I return it here? If so, we can remove this check.
https://github.com/kaiachain/kaia/pull/149/files/870a8b4894fbaa626fb1c4b51862eeb38b8ddf73#diff-0f2a03d410f3b42fb5cc232db0be82b42a741b952aa3999887370ef91c1eca58R304